### PR TITLE
Update dataset_config.py

### PR DESCRIPTION
**Pull request by naman0210**

### CODE REVIEW

File: pysaliency/dataset_config.py [hunk 1]
```diff
@@ -52,6 +54,115 @@ def apply_dataset_filter_config(stimuli, fixations, filter_config):
     return filter_fn(stimuli, fixations, **filter_config['parameters'])
 
 
+def filter_scanpaths_by_attribute(scanpaths: FixationTrains, whitelist: dict=None, blacklist: dict=None):
+    """Filter Scanpaths by values of scanpath attributes (fixation_trains.scanpath_attributes), the dictionary can have only one attribute"""
+    
+    assert (whitelist is None and blacklist is not None) or (whitelist is not None and blacklist is None)
+    if whitelist is not None:
+        assert(len(whitelist)==1)
```
Comment (matthias-k) [inline pysaliency/dataset_config.py:62]: my suggestion last friday was to only implement for now functions
```
def filter_scanpaths_by_attribute(scanpaths, attribute_name, attribute_value, invert_match=False):
    ....
```

They  would be sufficiently convenient for many applications and are much easier to write. What you already started to implement would be `filter_scanpaths_by_attributes` (note the trailing `s`), which then could simply call `filter_scanpaths_by_attribute` repeated times, or do a more efficient filter. But for now let's only for for the simple version that I just outlined. It will be easier to read and to use

File: pysaliency/dataset_config.py [hunk 2]
```diff
@@ -52,6 +54,115 @@ def apply_dataset_filter_config(stimuli, fixations, filter_config):
     return filter_fn(stimuli, fixations, **filter_config['parameters'])
 
 
+def filter_scanpaths_by_attribute(scanpaths: FixationTrains, whitelist: dict=None, blacklist: dict=None):
```
Comment (matthias-k) [inline pysaliency/dataset_config.py:57]: The filter functions themselves should go into `filter_datasets.py`. The present file, `dataset_config` should only use them to extend `apply_dataset_filter_config` (which is still missing from this PR).

**Review by matthias-k — CHANGES_REQUESTED** [review]
Can you please also add some very simple unittests? There is already `test_filter_datasets.py` including fixtures for scanpaths. The current fixtures don't include attributes, but check out `test_datasets.py` for fixtures with different attributes and also utility functions for comparing fixation objects `compare_fixations` and `compare_scanpaths`.

File: pysaliency/dataset_config.py (hunk 1 above)
Comment (matthias-k) [reply pysaliency/dataset_config.py:62]: `invert_match` is intended to choose between filtering positives or negatives

File: pysaliency/filter_datasets.py [hunk 3]
```diff
@@ -232,3 +232,50 @@ def filter_stimuli_by_size(stimuli, fixations, size=None, sizes=None):
     indices = [i for i in range(len(stimuli)) if stimuli.sizes[i] in sizes]
 
     return create_subset(stimuli, fixations, indices)
+
+
+def filter_scanpaths_by_attribute(scanpaths:  FixationTrains, attribute_name, attribute_value, invert_match=False):
+    """Filter Scanpaths by values of scanpath attribute (fixation_trains.scanpath_attributes)"""
+    
+    mask = np.array([element == attribute_value for element in getattr(scanpaths, attribute_name)])
```
Comment (matthias-k) [inline pysaliency/filter_datasets.py:240]: shouldn't this also work and be much more concise?
```suggestion
    mask = np.asarray(getattr(scanpaths, attribute_name)) == attribute_value
```

File: pysaliency/filter_datasets.py [hunk 4]
```diff
@@ -232,3 +232,50 @@ def filter_stimuli_by_size(stimuli, fixations, size=None, sizes=None):
...
+    if invert_match is True:
+        mask = ~mask
+    indices = list(np.nonzero(mask)[0])
```
Comment (matthias-k) [inline pysaliency/filter_datasets.py:243]: Do you need indices here? I thought the mask directly should work in filter_fixation_trains? But maybe I'm mistaken. It should definitely work for `Fixations` though

File: pysaliency/filter_datasets.py [hunk 5]
```diff
@@ -232,3 +232,50 @@ def filter_stimuli_by_size(stimuli, fixations, size=None, sizes=None):
...
+
+
+def filter_fixations_by_attribute(fixations: Fixations, attribute_name, attribute_value, invert_match=False):
+    """Filter Fixations by values of attribute (fixations.__attributes__)"""
+
+    mask = np.array([element == attribute_value for element in getattr(fixations, attribute_name)])
+    if invert_match is True:
+        mask = ~mask
+    indices = list(np.nonzero(mask)[0])
+    return fixations.filter(indices)
```
Comment (matthias-k) [inline pysaliency/filter_datasets.py:254]: see above:

```suggestion
    return fixations[mask]
```

File: tests/test_datasets.py [hunk 6]
```diff
@@ -747,5 +747,39 @@ def test_scanpaths_from_fixations(fixation_indices):
...
 
+invert_param_values = [True, False]
+@pytest.mark.parametrize('attribute_name , attribute_value', [('dva',1), ('dva',4), ('dva',15), ('some_strings','a'), ('some_strings','n'), ('some_strings','q')])
+@pytest.mark.parametrize('invert_match', invert_param_values)
+def test_filter_stimuli_by_attribute(file_stimuli_with_attributes, fixation_trains, attribute_name, attribute_value, invert_match):
+
+    fixations = fixation_trains[np.arange(len(fixation_trains))]
+    # Access the attribute using the attributes dictionary
+    attribute_data = file_stimuli_with_attributes.attributes[attribute_name]
+    mask = np.array([element == attribute_value for element in attribute_data])
```
Comment (matthias-k) [inline tests/test_datasets.py:758]: It's nice that you're testing multiple values via parametrize, but that means that you're essentially writing a regression test against the current implementation, which is copied here. I think it might be better to test only one or two values, and for those values handcode the correct indices (there are not that many), to check that it's actually the right ones. Right now, in theory your code could always match everything and the test would pass

File: tests/test_datasets.py [hunk 7]
```diff
@@ -747,5 +747,39 @@ def test_scanpaths_from_fixations(fixation_indices):
...
+    assert len(sub_stimuli) == len(stimulus_indices)
+    np.testing.assert_array_equal(sub_fixations.x, fixations.x[np.isin(fixations.n, stimulus_indices)])  
+
+
+invert_param_values = [True, False]
+@pytest.mark.parametrize('attribute_name , attribute_value', [('subjects',0), ('subjects',1), ('subjects',100), ('x',1), ('x',19), ('y',10), ('y',12), ('t',100), ('t',500), ('t',10000)])
+@pytest.mark.parametrize('invert_match', invert_param_values)
+def test_filter_fixations_by_attribute(fixation_trains, attribute_name, attribute_value, invert_match):
+    fixations = fixation_trains[:]
+    mask = np.array([element == attribute_value for element in getattr(fixations, attribute_name)])
```
Comment (matthias-k) [inline tests/test_datasets.py:776]: see above, let's test only one reasonable attribute, but with manually computed indices which we know to be correct for the fixture

**Review by matthias-k — CHANGES_REQUESTED** [review]
Thanks! I added some more comments and sugestions.

File: pysaliency/filter_datasets.py (hunk 3 above)
Comment (naman0210) [reply pysaliency/filter_datasets.py:240]: yes it does work, I'll change it :)

File: pysaliency/filter_datasets.py (hunk 4 above)
Comment (naman0210) [reply pysaliency/filter_datasets.py:243]: yes it works for `Fixaions` but for `FixationTrains` we need the indices

File: pysaliency/filter_datasets.py [hunk 8]
```diff
@@ -232,3 +232,55 @@ def filter_stimuli_by_size(stimuli, fixations, size=None, sizes=None):
...
+
+
+def filter_scanpaths_by_attribute(scanpaths:  FixationTrains, attribute_name, attribute_value, invert_match=False):
+    """Filter Scanpaths by values of scanpath attribute (fixation_trains.scanpath_attributes)"""
+
+    mask = scanpaths.scanpath_attributes[attribute_name] == attribute_value
+    if invert_match is True:
+        mask = ~mask
+    if mask.ndim>1:
+        mask = np.all(mask, axis=1)
```
Comment (matthias-k) [inline pysaliency/filter_datasets.py:244]: wouldn't this need to come before the `invert_match`? Right now, when `invert_match=False`, you're requiring _all_ entries to be different from the value.

File: pysaliency/filter_datasets.py [hunk 9]
```diff
@@ -232,3 +232,55 @@ def filter_stimuli_by_size(stimuli, fixations, size=None, sizes=None):
...
+
+
+def filter_fixations_by_attribute(fixations: Fixations, attribute_name, attribute_value, invert_match=False):
+    """Filter Fixations by values of attribute (fixations.__attributes__)"""
+
+    mask = np.asarray(getattr(fixations, attribute_name)) == attribute_value
+    if invert_match is True:
+        mask = ~mask
+    if mask.ndim>1:
+        mask = np.all(mask, axis=1)
```
Comment (matthias-k) [inline pysaliency/filter_datasets.py:256]: see above

File: pysaliency/filter_datasets.py [hunk 10]
```diff
@@ -232,3 +232,55 @@ def filter_stimuli_by_size(stimuli, fixations, size=None, sizes=None):
...
+    return fixations[mask]
+
+def filter_stimuli_by_attribute(stimuli: Stimuli, fixations: Fixations, attribute_name, attribute_value, invert_match=False):
+    """Filter stimuli by values of attribute (stimuli.attributes)"""
+
+    mask = np.asarray(stimuli.attributes[attribute_name]) == attribute_value
+    if invert_match is True:
+        mask = ~mask
+    if mask.ndim>1:
+        mask = np.all(mask, axis=1)
```
Comment (matthias-k) [inline pysaliency/filter_datasets.py:267]: see above

File: pysaliency/filter_datasets.py [hunk 11]
```diff
@@ -232,3 +232,55 @@ def filter_stimuli_by_size(stimuli, fixations, size=None, sizes=None):
...
+    return create_subset(stimuli, fixations, indices)
+
+
+def filter_scanpaths_by_lengths(scanpaths: FixationTrains, intervals: list):
+    """Filter Scanpaths by number of fixations"""
+    
+    intervals = _check_intervals(intervals, type=int)
+    mask = np.zeros(len(scanpaths.train_lengths), dtype=bool)
+    for n1, n2 in intervals:
+        temp_mask = np.logical_and(scanpaths.train_lengths>=n1,scanpaths.train_lengths<=n2)
```
Comment (matthias-k) [inline pysaliency/filter_datasets.py:279]: some suggestions:
1. let's use more verbose names than `n1, n2`
2. I always used the end as exclusive, to make it consistent with python indices like `array[start:end]`
3. I added some blanks to make it PEP8 compatible.
All together:
```suggestion
    for start, end in intervals:
        temp_mask = np.logical_and(scanpaths.train_lengths >= start, scanpaths.train_lengths < end)
```

File: tests/test_datasets.py [hunk 12]
```diff
@@ -747,5 +749,79 @@ def test_scanpaths_from_fixations(fixation_indices):
     compare_fixations(sub_fixations, new_sub_fixations, crop_length=True)
 
 
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('dva',1,False), ('some_strings','n',True), ('some_strings','b',True)])
+def test_filter_stimuli_by_attribute(file_stimuli_with_attributes, fixation_trains, attribute_name, attribute_value, invert_match):
+    new_fixations = pysaliency.Fixations.concatenate((fixation_trains, fixation_trains, fixation_trains[:2]))
+    new_fixations = new_fixations[:]
```
Comment (matthias-k) [inline tests/test_datasets.py:755]: why is this needed?

File: tests/test_datasets.py [hunk 13]
```diff
@@ -747,5 +749,79 @@ def test_scanpaths_from_fixations(fixation_indices):
...
+    _stimuli, _fixations = filter_stimuli_by_attribute(file_stimuli_with_attributes, new_fixations, attribute_name, attribute_value, invert_match)
```
Comment (matthias-k) [inline tests/test_datasets.py:756]: please use more verbose names, e.g.
```
    filtered_stimuli, filtered_fixations = ...
```

File: tests/test_datasets.py [hunk 14]
```diff
@@ -747,5 +749,79 @@ def test_scanpaths_from_fixations(fixation_indices):
...
+        compare_fixations(_fixations,_fixations2)
+        assert list(_stimuli.stimulus_ids) == file_stimuli_with_attributes.stimulus_ids[[1]]
+    if attribute_name == 'some_strings' and attribute_value == 'n' and invert_match is True:
+        _fixations2 = new_fixations[list(range(0,18))]
+        compare_fixations(_fixations,_fixations2)
+        assert list(_stimuli.stimulus_ids) == file_stimuli_with_attributes.stimulus_ids[list(range(0,13))+list(range(14,18))]
+    if attribute_name == 'some_strings' and attribute_value == 'b' and invert_match is True:
+        _fixations2 = new_fixations[[0,1,2,3,4,8,9,10,11,12,16,17]]
+        compare_fixations(_fixations,_fixations2)
+        assert list(_stimuli.stimulus_ids) == file_stimuli_with_attributes.stimulus_ids[list(range(0,1))+list(range(2,18))]
```
Comment (matthias-k) [inline tests/test_datasets.py:768]: if you need to write the test like this, don't use `parametrize`. Simply duplicate the test four times. That's much easier to read.

File: tests/test_datasets.py [hunk 15]
```diff
@@ -747,5 +749,79 @@ def test_scanpaths_from_fixations(fixation_indices):
...
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('subjects',0, True),('some_attribute',2, False), ('some_attribute',3, False),  ('some_attribute',4, False)])
+def test_filter_fixations_by_attribute(fixation_trains, attribute_name, attribute_value, invert_match):
+    fixations = fixation_trains[:]
+    _fixations1 = filter_fixations_by_attribute(fixations, attribute_name, attribute_value, invert_match)
```
Comment (matthias-k) [inline tests/test_datasets.py:774]: as above: please use more verbose names. Here I would suggest
```
filtered_fixations = filter_fixations_by_attribute(...)
expected_fixations = fixations[[3, 4, 5, 6, 7]]
```

File: tests/test_datasets.py [hunk 16]
```diff
@@ -747,5 +749,79 @@ def test_scanpaths_from_fixations(fixation_indices):
...
+        compare_fixations(_fixations1,_fixations2)
+    if attribute_name == 'some_attribute' and attribute_value == 4 and invert_match is False:
+        _fixations2 = fixations[[4]]
+        compare_fixations(_fixations1,_fixations2)
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('task',0, False), ('multi_dim_attribute',[2,3],False), ('multi_dim_attribute',[4,5.5],False)])
+def test_filter_scanpaths_by_attribute(fixation_trains, attribute_name, attribute_value, invert_match):
+    scanpaths = fixation_trains
+    _scanpaths1 = filter_scanpaths_by_attribute(scanpaths, attribute_name, attribute_value, invert_match)
```
Comment (matthias-k) [inline tests/test_datasets.py:792]: as above: dare to use verbose names ;).

File: tests/test_datasets.py (hunk 14 above)
Comment (matthias-k) [reply tests/test_datasets.py:768]: then you can also give better names, e.g.
```
def test_filter_stimuli_by_attribute: ...
def test_filter_stimuli_by_attribute_invert_match: ...
def test_filter_stimuli_by_attribute_string_value: ...
```

File: tests/test_datasets.py [hunk 17]
```diff
@@ -747,5 +749,79 @@ def test_scanpaths_from_fixations(fixation_indices):
     compare_fixations(sub_fixations, new_sub_fixations, crop_length=True)
 
 
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('dva',1,False), ('some_strings','n',True), ('some_strings','b',True)])
+def test_filter_stimuli_by_attribute(file_stimuli_with_attributes, fixation_trains, attribute_name, attribute_value, invert_match):
```
Comment (matthias-k) [inline tests/test_datasets.py:753]: These tests should go into `tests/test_filter_datasets.py`, because the functions are from `filter_datasets.py`. You can import `compare_fixations` etc from `test_datasets.py`, and you can also extend the fixturesin `filter_datasets.py` to have the additional attributes etc from `test_datasets.py`.

File: tests/test_datasets.py [hunk 18]
```diff
@@ -747,5 +749,79 @@ def test_scanpaths_from_fixations(fixation_indices):
...
+    new_fixations = pysaliency.Fixations.concatenate((fixation_trains, fixation_trains, fixation_trains[:2]))
+    new_fixations = new_fixations[:]
+    _stimuli, _fixations = filter_stimuli_by_attribute(file_stimuli_with_attributes, new_fixations, attribute_name, attribute_value, invert_match)
+    if attribute_name == 'dva' and attribute_value == 1 and invert_match is False:
+        _fixations2 = new_fixations[[5,6,7,13,14,15]]
```
Comment (matthias-k) [inline tests/test_datasets.py:758]: much easier and less error prone:
```
expected_stimuli, expected_fixations = create_subset(stimuli, fixations, [[1]])
```

File: tests/test_datasets.py [hunk 19]
```diff
@@ -747,5 +749,79 @@ def test_scanpaths_from_fixations(fixation_indices):
...
+        assert list(_stimuli.stimulus_ids) == file_stimuli_with_attributes.stimulus_ids[list(range(0,1))+list(range(2,18))]
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('subjects',0, True),('some_attribute',2, False), ('some_attribute',3, False),  ('some_attribute',4, False)])
+def test_filter_fixations_by_attribute(fixation_trains, attribute_name, attribute_value, invert_match):
+    fixations = fixation_trains[:]
+    _fixations1 = filter_fixations_by_attribute(fixations, attribute_name, attribute_value, invert_match)
+    if attribute_name == 'subjects' and attribute_value == 0 and invert_match is True:
+        _fixations2 = fixations[[3,4,5,6,7]]
+        compare_fixations(_fixations1,_fixations2)
```
Comment (matthias-k) [inline tests/test_datasets.py:777]: blanks are missing quite often:
```suggestion
        compare_fixations(_fixations1, _fixations2)
```

**Review by matthias-k — COMMENTED** [review]
I made a bunch of new comments. Also, I think the new filters are not yet added to `dataset_config.py`, right? That should be fairly straightforward and we need it to use the filter functions in our config files. Could you please add that as well?

Let me know if anything is unclear.

File: pysaliency/filter_datasets.py (hunk 8 above)
Comment (naman0210) [reply pysaliency/filter_datasets.py:244]: I am not sure what you mean by, requiring all entries to be difference from the value. In case of multi dim attribute I want value in each dimension to be equal, it can work before the invert_match if condition as well.

File: tests/test_datasets.py (hunk 17 above)
Comment (naman0210) [reply tests/test_datasets.py:753]: I am not sure if I get you on the " extend the fixturesin filter_datasets.py to have the additional attributes etc from test_datasets.py" part?

File: tests/test_datasets.py (hunk 12 above)
Comment (naman0210) [reply tests/test_datasets.py:755]: I just wanted to make it a bit more complicated, I'll remove that line :)

File: pysaliency/filter_datasets.py (hunk 4 above)
Comment (matthias-k) [reply pysaliency/filter_datasets.py]: are you sure? I just tested `scanpaths.filter_fixation_trains(scanpaths.train_ns==0)` and it worked well.

File: pysaliency/filter_datasets.py (hunk 8 above)
Comment (matthias-k) [reply pysaliency/filter_datasets.py:244]: no, it can't :). Let's assume one entry differs, then it shouldn't match (which it doesn't). But that also means that it should be a match with `invert_match=True`. But that's not the case:

1. original mask for that attribute value would be e.g. `[True, True, False, True]`
2. since `invert_match=True`, the mask is inverted to `[False, False, True, False]`
3. now `np.all` returns `False`, but it would need to return `True`.

If you do the accumulation before the `invert_match`, `np.all` would have worked (`[True, True, False, True] -> False -> True). After the `invert_match`, you would need to use `np.any` for `invert_match=False`. This is basically an application of `not (A and B) == (not A) or (not B)`

File: tests/test_datasets.py (hunk 12 above)
Comment (matthias-k) [reply tests/test_datasets.py:755]: I see :D.

File: tests/test_datasets.py (hunk 17 above)
Comment (matthias-k) [reply tests/test_datasets.py:753]: the fixtures `Stimuli` and `FixationTrains` in `test_datasets` come with attributes. The ones in `test_filter_datasets` don't have any attributes because I copied them over before I introduced attributes. But since you need the fixtures to have attributes to be able to apply the filter functions, you'll need to make sure that the fixtures in `test_filter_datasets` have attributes too.

File: pysaliency/filter_datasets.py (hunk 8 above)
Comment (naman0210) [reply pysaliency/filter_datasets.py:244]: ahhh okay now I get it. I will move it before the `invert_match` if condition.

File: tests/test_filter_datasets.py [hunk 20]
```diff
@@ -294,3 +341,109 @@ def test_stratified_crossval_splits_multiple_attributes(many_stimuli, crossval_f
     assert sum(len(f.x) for f in train_fixations) == (crossval_folds - val_folds - test_folds) * len(fixations.x)
 
     assert len(train_stimuli) == crossval_folds
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('dva', 1, False)])
+def test_filter_stimuli_by_attribute_dva(file_stimuli_with_attributes, fixation_trains, attribute_name, attribute_value, invert_match):
+    fixations = fixation_trains[:]
+    filtered_stimuli, filtered_fixations = filter_stimuli_by_attribute(file_stimuli_with_attributes, fixations, attribute_name, attribute_value, invert_match)
+    if attribute_name == 'dva' and attribute_value == 1 and invert_match is False:
```
Comment (matthias-k) [inline tests/test_filter_datasets.py:350]: why do you still test those values? The condition will always be true, right? You can just remove the if-clause and always run the next lines

File: tests/test_filter_datasets.py [hunk 21]
```diff
@@ -294,3 +341,109 @@ def test_stratified_crossval_splits_multiple_attributes(many_stimuli, crossval_f
...
+        compare_fixations(filtered_fixations, expected_fixations)
+        assert list(filtered_stimuli.stimulus_ids) == file_stimuli_with_attributes.stimulus_ids[inds]
+        assert_stimuli_equal(filtered_stimuli, expected_stimuli)
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('some_strings', 'n', True)])
+def test_filter_stimuli_by_attribute_some_strings_invert_match(file_stimuli_with_attributes, fixation_trains, attribute_name, attribute_value, invert_match):
+    fixations = fixation_trains[:]
+    filtered_stimuli, filtered_fixations = filter_stimuli_by_attribute(file_stimuli_with_attributes, fixations, attribute_name, attribute_value, invert_match)
+    if attribute_name == 'some_strings' and attribute_value == 'n' and invert_match is True:
```
Comment (matthias-k) [inline tests/test_filter_datasets.py:362]: see above

File: tests/test_filter_datasets.py [hunk 22]
```diff
@@ -294,3 +341,109 @@ def test_stratified_crossval_splits_multiple_attributes(many_stimuli, crossval_f
...
+        compare_fixations(filtered_fixations, expected_fixations)
+        assert list(filtered_stimuli.stimulus_ids) == file_stimuli_with_attributes.stimulus_ids[inds]
+        assert_stimuli_equal(filtered_stimuli, expected_stimuli)
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('subjects', 0, True)])
+def test_filter_fixations_by_attribute_subject_invert_match(fixation_trains, attribute_name, attribute_value, invert_match):
+    fixations = fixation_trains[:]
+    filtered_fixations = filter_fixations_by_attribute(fixations, attribute_name, attribute_value, invert_match)
+    if attribute_name == 'subjects' and attribute_value == 0 and invert_match is True:
```
Comment (matthias-k) [inline tests/test_filter_datasets.py:374]: see above

File: tests/test_filter_datasets.py [hunk 23]
```diff
@@ -294,3 +341,109 @@ def test_stratified_crossval_splits_multiple_attributes(many_stimuli, crossval_f
...
+        inds = [3, 4, 5, 6, 7]
+        expected_fixations = fixations[inds]
+        compare_fixations(filtered_fixations, expected_fixations)
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('some_attribute', 2, False)])
+def test_filter_fixations_by_attribute_some_attribute(fixation_trains, attribute_name, attribute_value, invert_match):
+    fixations = fixation_trains[:]
+    filtered_fixations = filter_fixations_by_attribute(fixations, attribute_name, attribute_value, invert_match)
+    if attribute_name == 'some_attribute' and attribute_value == 2 and invert_match is False:
```
Comment (matthias-k) [inline tests/test_filter_datasets.py:384]: see above

File: tests/test_filter_datasets.py [hunk 24]
```diff
@@ -294,3 +341,109 @@ def test_stratified_crossval_splits_multiple_attributes(many_stimuli, crossval_f
     assert sum(len(f.x) for f in train_fixations) == (crossval_folds - val_folds - test_folds) * len(fixations.x)
 
     assert len(train_stimuli) == crossval_folds
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('dva', 1, False)])
+def test_filter_stimuli_by_attribute_dva(file_stimuli_with_attributes, fixation_trains, attribute_name, attribute_value, invert_match):
+    fixations = fixation_trains[:]
```
Comment (matthias-k) [inline tests/test_filter_datasets.py:348]: why is this needed?

Comment (matthias-k) [reply tests/test_filter_datasets.py:348]: ah, you want to use the name `fixations`? Then I see

File: tests/test_filter_datasets.py [hunk 25]
```diff
@@ -294,3 +341,109 @@ def test_stratified_crossval_splits_multiple_attributes(many_stimuli, crossval_f
...
+    filtered_stimuli, filtered_fixations = filter_stimuli_by_attribute(file_stimuli_with_attributes, fixations, attribute_name, attribute_value, invert_match)
+    if attribute_name == 'dva' and attribute_value == 1 and invert_match is False:
+        inds = [1]
+        expected_stimuli, expected_fixations = create_subset(file_stimuli_with_attributes, fixations, inds)
+        compare_fixations(filtered_fixations, expected_fixations)
+        assert list(filtered_stimuli.stimulus_ids) == file_stimuli_with_attributes.stimulus_ids[inds]
+        assert_stimuli_equal(filtered_stimuli, expected_stimuli)
```
Comment (matthias-k) [inline tests/test_filter_datasets.py:355]: why do you need the first assertion? Isn't `assert_stimuli_equal` already taking care of that?

File: tests/test_filter_datasets.py [hunk 26]
```diff
@@ -294,3 +341,109 @@ def test_stratified_crossval_splits_multiple_attributes(many_stimuli, crossval_f
...
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('some_strings', 'n', True)])
+def test_filter_stimuli_by_attribute_some_strings_invert_match(file_stimuli_with_attributes, fixation_trains, attribute_name, attribute_value, invert_match):
+    fixations = fixation_trains[:]
+    filtered_stimuli, filtered_fixations = filter_stimuli_by_attribute(file_stimuli_with_attributes, fixations, attribute_name, attribute_value, invert_match)
+    if attribute_name == 'some_strings' and attribute_value == 'n' and invert_match is True:
+        inds = list(range(0, 13))+list(range(14, 18))
```
Comment (matthias-k) [inline tests/test_filter_datasets.py:363]: sorry, I got really used to PEP8 ;). I like that it makes the code a bit less dense.

```suggestion
        inds = list(range(0, 13)) + list(range(14, 18))
```

File: tests/test_filter_datasets.py [hunk 27]
```diff
@@ -294,3 +341,109 @@ def test_stratified_crossval_splits_multiple_attributes(many_stimuli, crossval_f
     assert sum(len(f.x) for f in train_fixations) == (crossval_folds - val_folds - test_folds) * len(fixations.x)
 
     assert len(train_stimuli) == crossval_folds
+
+
+@pytest.mark.parametrize('attribute_name , attribute_value, invert_match', [('dva', 1, False)])
```
Comment (matthias-k) [inline tests/test_filter_datasets.py:346]: if you pass only one set of parameters, there is no need to use `parametrize`. Simply use the values directly in the test.

File: tests/test_filter_datasets.py [hunk 28]
```diff
@@ -294,3 +341,109 @@ def test_stratified_crossval_splits_multiple_attributes(many_stimuli, crossval_f
...
+def test_filter_scanpaths_by_attribute_multi_dim_attribute_invert_match(fixation_trains, attribute_name, attribute_value, invert_match):
+    scanpaths = fixation_trains
+    filtered_scanpaths = filter_scanpaths_by_attribute(scanpaths, attribute_name, attribute_value, invert_match)
+    if attribute_name == 'multi_dim_attribute' and attribute_value == [0, 1] and invert_match is True:
+        inds = [1, 2]
+        expected_scanpaths = scanpaths.filter_fixation_trains(inds)
+        compare_scanpaths(filtered_scanpaths, expected_scanpaths)
+
+
+@pytest.mark.parametrize('intervals', [([(1, 2), (2, 3)]), ([(2, 3), (3, 4)]), ([(2)]), ([(3)])])
```
Comment (matthias-k) [inline tests/test_filter_datasets.py:430]: here the parameters and tests are sufficiently small, so here I see the point of parametrize.

**Review by matthias-k — COMMENTED** [review]
some more style and clarity comments :).

File: tests/test_datasets.py (hunk 17 above)
Comment (naman0210) [reply tests/test_datasets.py]: fixed

File: tests/test_datasets.py (hunk 18 above)
Comment (naman0210) [reply tests/test_datasets.py]: fixed

File: tests/test_filter_datasets.py (hunk 25 above)
Comment (naman0210) [reply tests/test_filter_datasets.py:355]: ohh I was using this before I started using the create_subset function in the test itself. But now since the create_subset is already tested I will remove this line.

File: tests/test_filter_datasets.py (hunk 24 above)
Comment (naman0210) [reply tests/test_filter_datasets.py:348]: no its not about the name, when you gave the skeleton of these functions, you wrote that for the function `filter_stimuli_by_attribute` the fixations argument should be of the class `Fixations`

File: tests/test_filter_datasets.py (hunk 28 above)
Comment (naman0210) [reply tests/test_filter_datasets.py:430]: okay

File: tests/test_filter_datasets.py (hunk 27 above)
Comment (naman0210) [reply tests/test_filter_datasets.py:346]: that is a personal choice. It just makes it more easily readable for me xD. But I can easily change that

File: tests/test_filter_datasets.py (hunk 24 above)
Comment (matthias-k) [reply tests/test_filter_datasets.py:348]: ah :). But `FixationTrains` is a subclass of `Fixations`, hence fixations is already a `Fixations` instance :). However I agree that sometimes it's good to make sure that it's not more than `Fixations`. Here, it shouldn't matter

File: tests/test_filter_datasets.py (hunk 20 above)
Comment (naman0210) [reply tests/test_filter_datasets.py:350]: yes will remove them

File: tests/test_filter_datasets.py (hunk 24 above)
Comment (naman0210) [reply tests/test_filter_datasets.py:348]: so should I change it to `FixationTrains` for this case?

File: tests/test_filter_datasets.py (hunk 27 above)
Comment (matthias-k) [reply tests/test_filter_datasets.py:346]: If it's on purpose, I'm fine with it :).

File: tests/test_filter_datasets.py (hunk 24 above)
Comment (matthias-k) [reply tests/test_filter_datasets.py:348]: you could, but I'm fine with both.

File: tests/test_filter_datasets.py (hunk 27 above)
Comment (naman0210) [reply tests/test_filter_datasets.py:346]: i changed it for now.

File: tests/test_filter_datasets.py (hunk 24 above)
Comment (naman0210) [reply tests/test_filter_datasets.py:348]: okay I'll leave it for `Fixations` for now.

File: tests/test_filter_datasets.py (hunk 22 above)
Comment (naman0210) [reply tests/test_filter_datasets.py]: fixed

File: tests/test_filter_datasets.py (hunk 21 above)
Comment (naman0210) [reply tests/test_filter_datasets.py]: fixed

File: tests/test_filter_datasets.py (hunk 23 above)
Comment (naman0210) [reply tests/test_filter_datasets.py]: fixed

File: tests/test_filter_datasets.py (hunk 26 above)
Comment (naman0210) [reply tests/test_filter_datasets.py]: fixed

File: tests/test_filter_datasets.py (hunk 25 above)
Comment (naman0210) [reply tests/test_filter_datasets.py]: fixed

**Review by matthias-k — APPROVED** [review]
looks good to merge now!
